src/pages/index.js

```diff
@@ -21,73 +21,229 @@ import {
+});
+
+// Form Validators
+const profileEditFormValidator = new FormValidator(options, profileEditForm);
+profileEditFormValidator.enableValidation();
+
+const addCardFormValidator = new FormValidator(options, addCardFormElement);
+addCardFormValidator.enableValidation();
+
+const avatarEditFormValidator = new FormValidator(
+  options,
+  document.querySelector("#change__profile_picture")
+);
+avatarEditFormValidator.enableValidation();
+
+// Image Popup
+const popupWithImage = new PopupWithImage({
+  popupSelector: ".preview__modal",
+});
+popupWithImage.setEventListeners();
+
+const deleteCardPopup = new PopupWithConfirmation({
+  popupSelector: "#delete__card-modal",
+  handleFormSubmit: (cardId, cardElement) => {
+    console.log(`Deleting card with cardId=${cardId}`);
+    // Check if card ID is from a hardcoded card
+    if (cardId.startsWith("card")) {
+      console.warn("Cannot delete hardcoded initial card");
+      cardElement.remove();
+      deleteCardPopup.close();
+      return;
+    }
+    // Proceed to delete card from server
+    deleteCardPopup.setDeleteState(true);
+    api
+      .deleteCard(cardId)
+      .then(() => {
+        console.log("Card successfully deleted");
+        cardElement.remove();
+        deleteCardPopup.close();
+      })
+      .catch((err) => {
+        console.error("Delete card error:", err);
+      })
+      .finally(() => {
+        // Reset the loading state regardless of success or error
+        deleteCardPopup.setDeleteState(false);
+      });
+  },
+});
+deleteCardPopup.setEventListeners();
+// Function to create card
+const createCard = (cardData) => {
+  return new Card(
+    cardData,
+    "#card-template", // The card template selector
+    () => {
+      // Handle image preview when the card image is clicked
+      popupWithImage.open(cardData);
+    },
+    (cardId, cardElement) => {
+      // Handle card deletion when the delete button is clicked
+      deleteCardPopup.open(cardId, cardElement);
+    },
+    handleLikeButton
+  ).generateCard();
+};
+//
 const cardSection = new Section(
   {
     items: initialCards,
     renderer: (cardData) => {
-      const cardInstance = new Card(cardData, "#card-template", () => {
-        popupWithImage.open(cardData);
-      });
-      const cardElement = cardInstance.generateCard();
+      const cardElement = createCard(cardData);
+      cardElement.id = cardData._id; // Ensure each card has a unique id
       cardSection.addItem(cardElement);
     },
   },
   ".cards__list"
 );
-cardSection.renderItems();
-//
+api
+  .getInitialCards()
+  .then((cards) => {
+    cardSection.renderItems(cards);
+  })
+  .catch((err) => {
+    console.error("Error fetching initial cards:", err);
+  });
+// Form Handlers
 const handleProfileFormSubmit = (data) => {
-  userInfo.setUserInfo({ name: data.name, job: data.job });
-  editProfilePopup.close();
+  editProfilePopup.setLoadingState(false); // Start loading state
```

Comment:
**Could be improved**

If it’s interesting for you here is how we can make a universal function for handling any submit. We can get rid of such duplicating as **loading effect**, **closing** and **catching errors**


```js

// You can make a universal function that accepts a request, popup instance and optional loading text  
function handleSubmit(request, popupInstance, loadingText = "Saving...") {

  // here we change the button text
  popupInstance.renderLoading(true, loadingText);
  request()
    .then(() => {
      // We need to close only in `then`
      popupInstance.close()
    })
      // we need to catch possible errors
      // console.error is used to handle errors if you don’t have any other ways for that
     .catch(console.error)
    // in `finally` we need to return the initial button text back in any case
    .finally(() => {
      popupInstance.renderLoading(false);
    });
}



```


You need to pass the **request** into the call of `handleSubmit` like that:



```js

// here is an example of the profile form handling
function handleProfileFormSubmit(inputValues) {
  // we create a function that returns a promise
  function makeRequest() {
    // `return` lets us use a promise chain `then, catch, finally` inside `handleSubmit`
    return api.editProfile(inputValues).then((userData) => {
      userInfo.setUserInfo(userData)
    });
  }
  // Here we call the function passing the request, popup instance and if we need some other loading text we can pass it as the 3rd argument
  handleSubmit(makeRequest, profilePopup);
}



```